flake-modules/updates/rust-analyzer.nix

```
@@ -0,0 +1,201 @@
+      {
+        kind = "list";
+        item = mkRustAnalyzerOptionType false "${property_name}-item" items;
+      }
+    else if type == "number" || type == "integer" then
+      {
+        kind = type;
+        inherit minimum maximum;
+      }
+    else if
+      lib.elem type [
+        "object"
+        "string"
+        "boolean"
+      ]
+    then
+      { kind = type; }
+    else
+      throw "Unhandled value in ${property_name}: ${lib.generators.toPretty { } property}";
+
+  mkRustAnalyzerOption =
+    property_name:
+    {
+      # List all possible values so that we are sure no new values are introduced
+      default,
+      markdownDescription,
+      enum ? null,
+      enumDescriptions ? null,
+      anyOf ? null,
+      minimum ? null,
+      maximum ? null,
+      items ? null,
+      # TODO: add this in the documentation ?
+      uniqueItems ? null,
+      type ? null,
+    }:
+    let
+      enumDesc =
+        values: descriptions:
+        let
+          valueDesc = builtins.map ({ fst, snd }: ''- ${fst}: ${snd}'') (
+            lib.lists.zipLists values descriptions
+          );
+        in
+        ''
+          ${markdownDescription}
+
+          Values:
+          ${builtins.concatStringsSep "\n" valueDesc}
+
+          default:
+          ```nix
+          ${lib.generators.toPretty { } default}
+          ```
+        '';
+    in
+    {
+      type = mkRustAnalyzerOptionType true property_name {
+        inherit
+          type
+          enum
+          minimum
+          maximum
+          items
+          anyOf
+          enumDescriptions
+          ;
+      };
+      description =
+        if
+          enum == null && (anyOf == null || builtins.all (subProp: !(lib.hasAttr "enum" subProp)) anyOf)
+        then
+          ''
+            ${markdownDescription}
+
+            default:
+            ```nix
+            ${lib.generators.toPretty { } default}
+            ```
+          ''
+        else if enum != null then
+          # We don't handle enum + anyOf types
+          assert anyOf == null;
+          enumDesc enum enumDescriptions
+        else
+          let
+            subEnums = lib.filter (lib.hasAttr "enum") anyOf;
+            subEnum =
+              # We only handle types with a single enum in an anyOf
+              assert lib.length subEnums == 1;
+              lib.head subEnums;
+          in
+          enumDesc subEnum.enum subEnum.enumDescriptions;
+    };
+
+  rustAnalyzerOptions = builtins.map (
+    v:
+    let
+      props = lib.attrsToList v.properties;
+      prop =
+        # We only handle items that have a single element
+        assert lib.length props == 1;
+        lib.head props;
+    in
+    {
+      "${prop.name}" = mkRustAnalyzerOption prop.name prop.value;
+    }
+  ) rustAnalyzerProperties;
+in
+writeText "rust-analyzer-options.nix" (
+  "# WARNING: this file is autogenerated\n"
+  + (lib.generators.toPretty { } (lib.foldl (a: b: a // b) { } rustAnalyzerOptions))
```

IMO this is more readable:

```suggestion
  + (lib.generators.toPretty { } (lib.foldl lib.updateAttrs { } rustAnalyzerOptions))
```

(Or `updateAttrsRecursive` if there's a chance we'd need that?)

---

Better yet, use [`mergeAttrsList`](https://noogle.dev/f/lib/mergeAttrsList):

```nix
(lib.mergeAttrsList rustAnalyzerOptions)
```

---

I tried to find such a function, but did not suceed! Thanks